pilot/pkg/proxy/envoy/v2/debounced_channel_test.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2_test
+
+import (
+	"testing"
+	"time"
+
+	v2 "istio.io/istio/pilot/pkg/proxy/envoy/v2"
+)
+
+func TestDebouncedChannelSimple(t *testing.T) {
+
+	dc := v2.NewDebouncedChannel(0, 0, false, func(accumulator v2.DebouncedEvent, value v2.DebouncedEvent) v2.DebouncedEvent {
+		return accumulator || value
+	})
+
+	expectNoEvent(t, dc)
+
+	dc.Enqueue(true)
+
+	expectEvent(t, dc, true)
+
+	expectNoEvent(t, dc)
+}
+
+func TestDebouncedChannelMultiple(t *testing.T) {
+
+	dc := v2.NewDebouncedChannel(0, 0, false, func(accumulator v2.DebouncedEvent, value v2.DebouncedEvent) v2.DebouncedEvent {
+		return accumulator || value
+	})
+
+	expectNoEvent(t, dc)
+
+	// first
+	dc.Enqueue(false)
+
+	// second
+	dc.Enqueue(true)
+	dc.Enqueue(false)
+
+	expectEvent(t, dc, false)
+	expectEvent(t, dc, true)
+
+	expectNoEvent(t, dc)
+}
+
+func TestDebouncedChannelMultipleWithDebounce(t *testing.T) {
+
+	dc := v2.NewDebouncedChannel(50*time.Millisecond, 1*time.Hour, false, func(accumulator v2.DebouncedEvent, value v2.DebouncedEvent) v2.DebouncedEvent {
+		return accumulator || value
+	})
+
+	expectNoEvent(t, dc)
+
+	dc.Enqueue(false)
+	dc.Enqueue(true)
+	dc.Enqueue(false)
+
+	expectNoEvent(t, dc)
```

Comment:
this seems racy. In prow our tests sometimes get heavily CPU throttled so I would worry this could cause failures. Maybe not very likely since it is 50ms though

Reply:
That's correct. What would you suggest? Currently it's a factor of 4 (50ms vs 200ms).